pkg/fioapp/layers.go

```diff
@@ -0,0 +1,224 @@
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"sort"
+)
+
+func GetManifestService(ctx context.Context, regClient internal.RegistryClient, repoRef reference.Named) (distribution.ManifestService, error) {
+	repo, err := regClient.GetRepository(ctx, repoRef)
+	if err != nil {
+		return nil, err
+	}
+	return repo.Manifests(ctx, nil)
+
+}
+
+func GetManifestList(ctx context.Context, manifestSvc distribution.ManifestService, digest digest.Digest) (*manifestlist.DeserializedManifestList, error) {
+	manifest, err := manifestSvc.Get(ctx, digest)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := manifest.(*manifestlist.DeserializedManifestList); !ok {
+		return nil, fmt.Errorf("invalid manifest type, expected manifestlist.DeserializedManifestList, got: %T", manifest)
+	}
+	return manifest.(*manifestlist.DeserializedManifestList), nil
+}
+
+func GetManifest(ctx context.Context, manifestSvc distribution.ManifestService, digest digest.Digest) (*schema2.DeserializedManifest, error) {
+	manifest, err := manifestSvc.Get(ctx, digest)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := manifest.(*schema2.DeserializedManifest); !ok {
+		return nil, fmt.Errorf("invalid manifest type, expected schema2.DeserializedManifest, got: %T", manifest)
+	}
+	return manifest.(*schema2.DeserializedManifest), nil
+}
+
+func GetAppLayers(ctx context.Context, services map[string]types.ServiceConfig) (map[string][]distribution.Descriptor, error) {
+	svcImages := make(map[string]string)
+	for svc, svcCfg := range services {
+		svcImages[svc] = svcCfg.Image
+	}
+	return GetAppLayersFromMap(ctx, svcImages)
+}
+
+func GetLayers(ctx context.Context, services map[string]interface{}) (map[string][]distribution.Descriptor, error) {
+	svcImages := make(map[string]string)
+	for svc, cfg := range services {
+		svcCfg := cfg.(map[string]interface{})
+		svcImages[svc] = svcCfg["image"].(string)
+	}
+	return GetAppLayersFromMap(ctx, svcImages)
+}
+
+func GetAppLayersFromMap(ctx context.Context, svcImages map[string]string) (map[string][]distribution.Descriptor, error) {
+	regClient := internal.NewRegistryClient()
+
+	// Get manifests per architecture and per image, for one architecture there should be one manifest for each service image
+	// One image might have two or more manifests per the same architecture, it's odd, but is true, see nginx's manifest list
+	// Thus, we have to created a map of maps, arch -> image/manifest-service -> manifest-digest, to avoid inclusion
+	// more then manifest per image and per architecture
+	archToManifestList := make(map[string]map[distribution.ManifestService]digest.Digest)
+	for _, image := range svcImages {
+		imageRef, err := reference.ParseNamed(image)
+		if err != nil {
+			return nil, err
+		}
+
+		imageManifestSvc, err := GetManifestService(ctx, regClient, imageRef)
+		if err != nil {
+			return nil, err
+		}
+
+		manifestList, err := GetManifestList(ctx, imageManifestSvc, imageRef.(reference.Canonical).Digest())
```

Comment:
> I don't think we should force multi-arch images.

I am not sure I understand this statement. We don't force anything here, just simply parsing images that were generated by CI jobs.

Reply:
we also allow people to use containers created outside of hub.foundries.io that aren't always multi-arch manifests. In the event a customer used a normal manifest, this could would fail.

Reply:
Indeed, somehow I didn't think about such use-case. I'll fix it.

Reply:
>Maybe there's some re-use there so we don't have to look everything up 2 times? e.g. PinServiceImages could return a struct with all these layers for you to consume

I actually started from that point and had considered it, but it turned out that adjusting to the code re-usage actually creates more code than the amount of the reused code. Also, such decoupling makes it easier to develop, test and debug.

Reply:
>we also allow people to use containers created outside of hub.foundries.io that aren't always multi-arch manifests. In the event a customer used a normal manifest, this could would fail.

I am not sure that we can support such use-case at all since the "normal" manifest ("application/vnd.docker.distribution.manifest.v2+json") does not include any information about platform/architecture while we group layers of all images by architecture. For example, if we have three images, two of them have manifests for ['amd64', 'arch64'], the third image has just "normal" manifest with no info on its architecture, so the question is, is it image for amd64 or arch64?

Reply:
This is how the pinning service works. It pins both a multi-arch image (hub.foundries.io/msul-dev01/app-05) that supports amd64 and arm64 as well as a single arch/manifest image docker.io/foundries/app-05 which supports just amd64.
I think we should fail in this case, unless there is just one arch/platform specified in factory-config.yml.
Also, I think we should pass the platform list specified in factory-config.yml from ci-scripts to compose-publish so the later can verify that each image supports a given list of architectures. In this case, if the list contains just one element then we can except images with a "normal" manifest.

```
 = Pinning service images...
| Pinning void-service(foundries/app-05:tag01)
| 
|   |->  docker.io/foundries/app-05@sha256:7736be43d0605bb0314a673ebfe6ed4dd95d40380c55ba6cb496db7dedde683b
| Pinning python-www(hub.foundries.io/msul-dev01/app-05:1108_79fd4c6)
|   | amd64, arm64
|   |->  hub.foundries.io/msul-dev01/app-05@sha256:ca456ca35da932f1494de0bdc8895a12c98c31b4d8cbe28b9c1d5a3f0d02ffea
```

Reply:
There's a way to get the platform. I think it comes from the config, but you can see it with something like:
```
$ skopeo inspect docker://doanac/jobserv:latest | jq .Architecture
"amd64"

$ skopeo inspect docker://arm32v7/ruby:latest | jq .Architecture
"arm"
```

Reply:
ok, I'll check it out.

I suppose we should pass a list of platforms specified in factory-config.yml to compose-publish anyway, otherwise we might end up publishing App that doesn't support these platforms.

Reply:
So, the following are logs for App with mixed type of images (app05 - python-www void-service - manifest)

```
| Pinning python-www(hub.foundries.io/msul-dev01/app-05:1110_6cca820)
|   | amd64, arm64
|   |->  hub.foundries.io/msul-dev01/app-05@sha256:ca456ca35da932f1494de0bdc8895a12c98c31b4d8cbe28b9c1d5a3f0d02ffea
| == Hashing services...
|    |-> void-service : 0eee99b8075d8b07458f412e56a302f0b7722405be683018b4492baa0fb91f73
|    |-> python-www : 3bcecb4fbbf1add598a96eaf774245d94787f362f3ef8e4d40939b01fd1c9abd
| = Getting app layers metadata...
|   |-> getting app layers for architecture: amd64
|   |-> exclude  arm64 architecture, some of the app images (1 images) don't have manifest for it
| = Posting app layers manifests...
|   |-> posting a layer manifest for architecture: amd64...OK |-> digest: 
```

https://ci.foundries.io/projects/msul-dev01/lmp/builds/1110/publish-compose-apps/artifacts/console.log

Reply:
Now, on my arm device I have subtle issue. aklite says that update was successful. One of the containers of my app-05 (python-www) is running, another one (void-service) not.
Both, fioctl and the UI reports that the update was successful https://app.foundries.io/factories/msul-dev01/devices/msul-rpi4-98755/.

Even the report events didn't catch the error because aklite gets containers status just after they started so the void-service didn't fail by the time the status was taken.

```
OSTree hash already installed, same as current
# Apps installed:
hub.foundries.io/msul-dev01/app-05@sha256:6b8b7ddac8a9fd39e3bc17ec3a48e31b6817d827f9adc1a7c4d4145fa76daf50
# Apps running:
app-05: ; state: 
	python-www: 3bcecb4fbbf1add598a96eaf774245d94787f362f3ef8e4d40939b01fd1c9abd; image: hub.foundries.io/msul-dev01/app-05@sha256:ca456ca35da932f1494de0bdc8895a12c98c31b4d8cbe28b9c1d5a3f0d02ffea; state: running; status: Up 46 minutes
	void-service: 0eee99b8075d8b07458f412e56a302f0b7722405be683018b4492baa0fb91f73; image: docker.io/foundries/app-05@sha256:7736be43d0605bb0314a673ebfe6ed4dd95d40380c55ba6cb496db7dedde683b; state: running; status: Up Less than a second
app-07: ; state: 
	nginx-01: 43eb18209da71e7501c06bb06b7a0bcaeedb86a3b98db03589bebba363ff2cde; image: hub.foundries.io/msul-dev01/nginx@sha256:67960643d8f50607a1df975581d29921bde4991e2b97bf3bc0c00b4fddd7b05b; state: running; status: Up 47 minutes
	nginx-02: b920517847f29f65b975e441751225e6028f330f591592c5a5eaa72386da65a4; image: docker.io/library/nginx@sha256:a97eb9ecc708c8aa715ccfb5e9338f5456e4b65575daf304f108301f3b497314; state: running; status: Up 47 minutes
	python-www: d27a8a19e10721f2edc974bcced2d0ba5558c728a8544a2bfcfeaa5561546389; image: hub.foundries.io/msul-dev01/app-07@sha256:b0150d88116219cbf46ebb5dc08d8a559c4f1ab2731a788628fc7375b2372cb0; state: running; status: Up 47 minutes
```

So, I think, it's really important to prevent it at the CI stage by failing the publish run and outputting meaningful error.

Reply:
i agree. but that's orthogonal to this, right? should we just get this merged and follow up with another PR focused on that?

Reply:
sure, I was just preparing input info for another task(s) related to compose-publish :).